homeassistant/helpers/entityfilter.py

```diff
@@ -0,0 +1,74 @@
+"""Helper class to implement include/exclude of entities and domains."""
+
+from homeassistant.const import (
+    CONF_ENTITIES, CONF_DOMAINS)
+from homeassistant.core import split_entity_id
+
+
+class EntityFilter():
+    """Class that implements an include/exclude filter for entities/domains."""
+
+    def __init__(self, include_dict, exclude_dict):
+        """Set up the filter."""
+        self.include_e = include_dict.get(CONF_ENTITIES, [])
+        self.include_d = include_dict.get(CONF_DOMAINS, [])
+        self.exclude_e = exclude_dict.get(CONF_ENTITIES, [])
+        self.exclude_d = exclude_dict.get(CONF_DOMAINS, [])
+
+        if self.exclude_e + self.exclude_d == []:
+            self.have_exclude = False
+        else:
+            self.have_exclude = True
+
+        if self.include_e + self.include_d == []:
+            self.have_include = False
+        else:
+            self.have_include = True
+
+    def check_entity(self, entity_id):
```

Comment:
I wonder if instead of making this a class, we should make it a function. And then to one up this, we could have the config validation return this function! If no config is passed in, config validation would return a function that always returns True.

Reply:
The rest of the comments make sense, and I'll get them fixed up.  I'm not sure what you mean by this one - sorry, I'm fairly new to Python.  If there's a similar setup (config validation returning a helper function) that you can point me at for an example, that would be great.

Also, would that allow multiple 'include/exclude' blocks in the same config?  I'm looking at adding this to mqtt_statestream to allow selecting what actually gets published, but also in a sub-config - I've got a working patch that allows publishing the MQTT autoconfig for entities as well as the states, but it really needs a different criteria for selection (really only makes sense for sensors and binary_sensors, but there could be exceptions).

I'm envisioning something like this:

```
mqtt_statestream:
  base_topic: homeassistant
  publish_attributes: true
  publish_timestamps: true
  include:
    domains:
      - light
      - sensor
      - binary_sensor
      - lock
  autoconfig:
    publish: true
    include:
      domains:
        - sensor
        - binary_sensor
```

Reply:
Okay, let's forget about the config validation part doing magic. That's for a future PR.

By returning just a function, you hide the sets and that means that you can do the checks at the beginning. Something like this

```python
def generate_filter(include_domains, include_entities,
                    exclude_domains, exclude_entities):
    
    …

    # Case 1 - no includes or excludes - pass all entities
    if not have_include and not have_exclude:
        return lambda entity_id: True

   # Case 2 - includes, no excludes - only include specified entities
   if self.have_include and not self.have_exclude:
        def filter(entity_id):
            return (entity_id in self.include_e or
                    domain in self.include_d)

        return filter

    etc…
```

Then use that like this:

```python
entity_filter = generate_filter(…)

if entity_filter('light.kitchen'):
    …
```

Reply:
Ah, gotcha!  No problem, I'll convert it tonight.  Thanks!